Improvements over actions available on the layer for QFieldCloud (#508)
opengisch/QField-docs: documentation/reference/qfieldcloud/system.en.md

---

# QFieldCloud System Documentation

The aim of this document is to provide an overview of the QFieldCloud system to understand the underlaying logic and technology.

## Entities and Concepts

### QGIS Project

> In QFieldCloud context a QGIS project refers to all data files that are required for properly functioning QGIS project.

Reviewer: I'm not sure I understand what this sentence should mean.

In the QFieldCloud context, a *project* refers to all data files that are required for a properly functioning QGIS project.

?

Author: ```suggestion
In the QFieldCloud context, a project refers to all data files that are required for a properly functioning QGIS project.
```

Read more about [QGIS projects](./projects.md).

### Layer action

> Each layer in the QGIS project can be configured with an "layer action".

Reviewer: .. a "layer action".

Author: ```suggestion
Each layer in the QGIS project can be configured with a "layer action".
```

The action determines how QFieldSync and QField should treat the layer.

There are the two actions that can be configured: cloud action and cable action.

> They work are applied in the QFieldCloud and traditional cable export context respectively.

Reviewer: They work in the QFieldCloud and traditional cable export context respectively.

Author: ```suggestion
They work in the QFieldCloud and traditional cable export context respectively.
```

In summary, with QFieldCloud:

- *Offline editing* means that a offline copy of the data will be generated by QFieldCloud and downloaded to QField.
The data from *Offline editing*-layers will all be stored as multiple layers within a single `data.gpkg`.
Whenever a modification is made on QField, a JSON structure called a *Change* (or Delta) will be generated only for the features that have been modified, only for the attributes or the geometry that have been modified.
On push, the original data source will be updated, including the database for database layers.
- *Directly access data source* is mainly used for service-based layers that are located on a internet accessible server.
The data in these layers are directly modified by QField.
Examples are WFS, WM(T)S-layers or layers coming from a database such as PostGIS layers.
If the layer is file-based it will be in read-only mode.
- *Remove from project* will simply remove the layer from the project (not package it for QField).

> From QFieldSync it will be possible to update a project already loaded on QFieldCloud.

Reviewer: With QFieldSync in QGIS it will be possible to update a project already loaded on QFieldCloud.

Author: ```suggestion
With QFieldSync in QGIS it will be possible to update a project already loaded on QFieldCloud.
```

If there are changes in the data or the structure of the layers, then the new data source files will be updated.

Note that changing the structure (or schema) of the layers such as adding, removing or renaming attributes, changing not null constraints, etc. might have effects.
If such changes are pushed to QFieldCloud, please make sure all QField users have pushed their changes in advance.
Otherwise it might be impossible for QFieldCloud to apply these Changes/Deltas from older version of the QGIS project.

#### Behaviour of QField

Depending on the actions set for each layer in QFieldSync for QFieldCloud or Cable export, QField will act as follows:

| Action                          | File based layer          | Service-based layer (e.g. WMS, database)   | Notes                                |
|---------------------------------|---------------------------|--------------------------------------------|--------------------------------------|
| Offline editing                 | Create and push a Change  | N/A                                        |                                      |

> | Directly access data source     | Layer is readonly         | Edit the online online database            |                                      |

Reviewer: Edit the online database

(there are two "onlines"

Author: ```suggestion
| Directly access data source     | Layer is readonly         | Edit the online  database                  |                                      |
```
